src/main/java/com/google/sps/ProductDetectionAPIImpl.java

```diff
@@ -0,0 +1,143 @@
+import java.util.List;
+
+/**
+ * Uses Cloud Vision API to detect product photo content.
+ */
+public class ProductDetectionAPIImpl implements ProductDetectionAPI {
+
+  public ProductDetectionData detectProductPhotoContent(byte[] imageBytes) 
+      throws PhotoDetectionException {
+    // Build Image object by converting bytes array.
+    ByteString byteString = ByteString.copyFrom(imageBytes);
+    Image image = Image.newBuilder().setContent(byteString).build();
+
+    // Build the image annotation request.
+    // {@code requestBuilder} holds the image and the features to detect from it.
+    Builder requestBuilder = AnnotateImageRequest.newBuilder();
+    
+    requestBuilder.setImage(image);
+
+    Feature labelDetectionFeature = 
+        Feature.newBuilder().setType(Feature.Type.LABEL_DETECTION).build();
+    requestBuilder.addFeatures(labelDetectionFeature);
+
+    Feature logoDetectionFeature =
+        Feature.newBuilder().setType(Feature.Type.LOGO_DETECTION).build();
+    requestBuilder.addFeatures(logoDetectionFeature);
+
+    Feature imagePropertiesDetectionFeature =
+        Feature.newBuilder().setType(Feature.Type.IMAGE_PROPERTIES).setMaxResults(5).build();
+    requestBuilder.addFeatures(imagePropertiesDetectionFeature);
+
+    List<AnnotateImageRequest> requests = new ArrayList<>();
+    requests.add(requestBuilder.build());
+
+    // Initialize client that will be used to send requests. This client only needs to be created
+    // once, and can be reused for multiple requests.
+    ImageAnnotatorClient imageAnnotatorClient;
+    try {
+      imageAnnotatorClient = ImageAnnotatorClient.create();
+    } catch (IOException exception) {
+      throw new PhotoDetectionException(
+          "Failed to create ImageAnnotatorClient.\n" + exception.getMessage(), exception);
+    }
+
+    // Perform detection on the image file.
+    BatchAnnotateImagesResponse response = imageAnnotatorClient.batchAnnotateImages(requests);
+    
+    List<AnnotateImageResponse> responses = response.getResponsesList();
+
+    List<String> labels = new ArrayList();
+    List<String> logos = new ArrayList();
+    List<String> colors = new ArrayList();
+
+    for (AnnotateImageResponse res : responses) {
+      if (res.hasError()) {
+        throw new PhotoDetectionException(res.getError().getMessage());
+      }
+
+      labels.addAll(getLabels(res));
+      logos.addAll(getLogos(res));
+      colors.addAll(getColors(res));
+    }
+
+    ProductDetectionData productDetectionData = new ProductDetectionData(labels, logos, colors);
+
+    // All requests have been completed, so call the "close" method on the client 
+    // to safely clean up any remaining background resources.
+    imageAnnotatorClient.close();
+
+    return productDetectionData;
+  }
+
+  private List<String> getLabels(AnnotateImageResponse res) {
+    List<String> labels = new ArrayList(); 
+        
+    for (EntityAnnotation annotation : res.getLabelAnnotationsList()) {
```

Comment:
This and following could be done with Java stream. e.g.

return  res.getLabelAnnotationsList().stream().map(a -> a.getDescription).collect(ImmutableList::toImmutableList);



https://winterbe.com/posts/2014/07/31/java8-stream-tutorial-examples/